File: 11-container-with-most-water/container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        left,right = 0, len(height) - 1

        max_water = 0

        while left < right:
            h = min(height[left],height[right])
            w = right - left
            area = h * w
            max_water= max(max_water,area)

            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        
        return max_water

        # SC -> O(1)
        # TC -> O(N)
        

File: 11-container-with-most-water/container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int mostWater = 0;
        int i = 0;
        int j = height.length -1;

        while(i < j){
            int l = j - i;
            int h = Math.min(height[i],height[j]);

            mostWater = Math.max(mostWater, h*l);

            if(height[i] < height[j]){
                i+=1;
            }else{
                j-=1;
            }
        }

        return mostWater;
    }
}

// SC -> O(1)
// TC -> O(N)